### PR TITLE
Fix bug #1109 drop down list modal event pump race.

### PR DESCRIPTION
**Pull request by LGM-Doyle**
tl;dr; A race condition between the EventPump and the ModalEventPump in the DropDownList code was fixed by adding a separate signal SelChangedWhileDroppedSignal.

This fixes issue #1109.

****The problem****

+ DropDownList uses a ModalEventPump to display the drop down list in ModalListPicker, while at the same time moving and restoring the selected row each Render() phase to render it in the drop down list anchor location.  
+ SidePanel::SystemRow displays a centered system name.  
+ SidePanel deletes and re-creates all of the rows immediately when SystemRow indicates a change in the selected row.  
+ The DropDownList and the DropDownList::ModalListPicker are both running PreRender on the same SystemRow in different EventPumps.  

A race condition happens which sometimes results in the SystemRow::Init() function running twice.  This creates an inconsistent SystemRow, which eventually leads to the crash.

Triggering this bug does not require the SidePanel to be touched in anyway.  It might require the SidePanel to have been visible at least once, but  I did not check this condition.

****The evidence****

I captured the stack inside of the SystemRow::Init() function showing that it had a) run unnecessarily a second time when m_intialized was already true and b) that it added a spurious control with a min width in Layout of 32767 !!!.  SystemRow while valiantly keeping the System name centered aliases the spurious min width with certain widths of SidePanels to create either ever larger magnitude positive/negative (m_ul_prerender, m_lr_prerender) pairs.  These are the large values that Geoff noticed that look uninitialized.  This is also what I assume was causing the crash to be very machine specific.

****A solution****

1. Remove the clamping of SystemRow to the parent size that was interacting badly with the DropDownList double render of the selected row.
2. Prevent the DropDownWnd correcting the size of the drop down box while the drop down box is visible.  This should prevent both DropDownList::PreRender() and ModalListPicker::PreRender() from racing to prerender the list box.
3. Make EndRun() virtual so that ModalListPicker can hide its list as part of its teardown within the ModalEventPump.
4. Split SelChangedSignal in to SelChangedWhileDroppedSignal and SelChangedSignal so that consumers can handle them appropriately.  I considered passing a parameter `is_dropped`, but chose two signals to emphasize that the SelChangedWhileDropped version is handled in the ModalEventPump and not the normal event pump.  The user could also use a single handler and detect status with Dropped().
5. Change the original SelChangedSignal to emit a final SelChanged when the drop down closes so that old code does not miss any selection changes events.
6. Changed the SidePanel code to separate the re-generation of the SystemRow so that it doesn't happen while the modal list is open.  

****Not perfect****

It introduces (at the least) the following quirk/bug.  If the system name drop down list is open and the names of known systems changes this will not be reflected in the systems in the list until the list is closed and re-opened.  A solution to that is to refactor DropDownList to not use  ModalEventPump.

### CODE REVIEW

File: UI/SidePanel.cpp [hunk 1]
```diff
@@ -3056,54 +3038,67 @@ void SidePanel::RefreshInPreRender() {
...
+void SidePanel::RefreshSystemNames() {
+    TemporaryPtr<const System> system = GetSystem(s_system_id);
+    // if no system object, there is nothing to populate with.  early abort.
+    if (!system)
+        return;
+
+    // Repopulate the system with all of the names of known systems, if it is closed.
+    // If it is open do not change the system names because it runs in a seperate ModalEventPump
+    // from the main UI.
+    if (!m_system_name->Dropped()) {
```
Comment (geoffthemedio) [inline UI/SidePanel.cpp:3050]: prefer
````
if (test)
    return
(non-indented code)
````
to
````
if (!test) {
    (bunch of indented code)
}
````

File: UI/SidePanel.cpp [hunk 2]
```diff
@@ -3056,54 +3038,67 @@ void SidePanel::RefreshInPreRender() {
...
+        std::map<std::string, int> system_map; //alphabetize Systems here
+        for (ObjectMap::const_iterator<System> sys_it = Objects().const_begin<System>();
+             sys_it != Objects().const_end<System>(); ++sys_it)
+        {
+            if (!sys_it->Name().empty() || sys_it->ID() == s_system_id) // skip rows for systems that aren't known to this client, except the selected system
+                system_map.insert(std::make_pair(sys_it->Name(), sys_it->ID()));
+        }
+        std::vector<GG::DropDownList::Row*> rows;
+        rows.reserve(system_map.size());
+        for (std::map< std::string, int>::iterator sys_it = system_map.begin(); sys_it != system_map.end(); ++sys_it)
```
Comment (geoffthemedio) [inline UI/SidePanel.cpp:3063]: break over line?

File: UI/SidePanel.cpp [hunk 3]
```diff
@@ -3056,54 +3038,67 @@ void SidePanel::RefreshInPreRender() {
...
+        {
+            int sys_id = sys_it->second;
+            rows.push_back(new SystemRow(sys_id));
+        }
+        m_system_name->Insert(rows, false);
+
+        // select in the list the currently-selected system
```
Comment (geoffthemedio) [inline UI/SidePanel.cpp:3070]: does inserting invalidate other row iterators? if no, can you select when adding instead of as a separate step? probably not, as the preexisting code did it this way...

**Comment by geoffthemedio** [comment]
reportedly helps, so merging...
